stdlib/Random/src/XoshiroSimd.jl

```diff
@@ -0,0 +1,344 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+module XoshiroSimd
+# Getting the xoroshiro RNG to reliably vectorize is somewhat of a hassle without Simd.jl.
+import ..Random: TaskLocal, rand, rand!, UnsafeView, Xoshiro, SamplerType, CloseOpen12, SamplerTrivial
+using Base: BitInteger_types
+
+# Vector-width. Influences random stream. We may want to tune this before merging.
+xoshiroWidth() = Val(8)
+# Simd threshold. Influences random stream. We may want to tune this before merging.
+simdThreshold(::Type{T}) where T = 2048
+simdThreshold(::Type{Bool}) = 4096
+
+@inline _rotl45(x::UInt64) = (x<<45)|(x>>19)
+@inline _shl17(x::UInt64) = x<<17
+@inline _rotl23(x::UInt64) = (x<<23)|(x>>41)
+@inline _plus(x::UInt64,y::UInt64) = x+y
+@inline _xor(x::UInt64,y::UInt64) = xor(x,y)
+@inline _and(x::UInt64, y::UInt64) = x & y
+@inline _or(x::UInt64, y::UInt64) = x | y
+@inline _lshr(x, y::Int32) = _lshr(x, y % Int64)
+@inline _lshr(x::UInt64, y::Int64) = Core.Intrinsics.llvmcall("""
+    %res = lshr i64 %0, %1
+    ret i64 %res
+    """,
+    UInt64,
+    Tuple{UInt64, Int64},
+    x, y)
+
+
+# required operations. These could be written more concise with `ntuple`, but the compiler
+# sometimes refuses to properly vectorize.
+for N in [1,2,4,8,16,32]
+    let
+        local code
+        code = """
+        %lshiftOp = shufflevector <1 x i64> <i64 45>, <1 x i64> undef, <$N x i32> zeroinitializer
+        %rshiftOp = shufflevector <1 x i64> <i64 19>, <1 x i64> undef, <$N x i32> zeroinitializer
+        %lshifted = shl <$N x i64> %0, %lshiftOp
+        %rshifted = lshr <$N x i64> %0, %rshiftOp
+        %res = or <$N x i64> %lshifted, %rshifted
+        ret <$N x i64> %res
+        """
+        @eval @inline _rotl45(x::NTuple{$N, VecElement{UInt64}}) = Core.Intrinsics.llvmcall($code,
+            NTuple{$N, VecElement{UInt64}},
```

Comment:
https://llvm.org/docs/LangRef.html#llvm-fshl-intrinsic
LLVM now has the "funnel shift left/right" intrinsics for rotates.

For example

```julia
using Base: llvmcall, VecElement
for N in [1,2,4,8,16,32]
    s = ntuple(_ -> Core.VecElement(45%UInt64), Val(N))
    op = "llvm.fshl.v$(N)i64"
    @eval @inline function _rotl45(x::NTuple{$N,VecElement{UInt64}})
        ccall($op, llvmcall, NTuple{$N,VecElement{UInt64}}, (NTuple{$N,VecElement{UInt64}},NTuple{$N,VecElement{UInt64}},NTuple{$N,VecElement{UInt64}}), x, x, $s)
    end
end
```
A little shorter, and as LLVM's documentation specifically says that rotate can be implemented by passing the same vector as the first two arguments, we can be confident that this'll reliably produce rotate instructions. E.g.:
```julia
julia> vxu = ntuple(_ -> VecElement(rand(UInt64)), Val(8));

julia> @code_native debuginfo=:none syntax=:intel _rotl45(vxu)
        .text
        vprolq  zmm0, zmm0, 45
        ret
        nop     dword ptr [rax + rax]
```
I didn't try the current definition, but I believe (last time I tried) that it was also optimized into a rotate.

Reply:
Using too long a vector naturally does exactly what we'd want it to:
```julia
julia> vxu = ntuple(_ -> VecElement(rand(UInt64)), Val(16));

julia> @cn _rotl45(vxu)
        .text
        vprolq  zmm0, zmm0, 45
        vprolq  zmm1, zmm1, 45
        ret
        nop
```
So `xoshiroWidth() = 8` should be fine. 512-bit SIMD architectures won't get as much out of order execution as others, but that's not the end of the world. That's what SMT is for...

Reply:
I just confirmed that the current code in the PR produces the same assembly on my machine.